File: backends/python-fastapi/stackverse_backend/routers/messages.py
from __future__ import annotations

from typing import Any
from uuid import uuid4

from fastapi import APIRouter, Request, Response
from psycopg import sql
from psycopg.errors import UniqueViolation

from ..api import (
    duplicate_message_conflict,
    log_message_event,
    message_snapshot,
    page_of,
    to_message_response,
    validate_message_input,
    where_clause,
)
from ..audit import record_audit
from ..auth import AdminCaller, OptionalCaller
from ..db import one, query, transaction
from ..etag import response_with_etag
from ..i18n import message_bundle, resolve_language
from ..problems import (
    NotFoundProblem,
    escape_like,
    first_param,
    parse_uuid,
    require_max_length,
    require_valid_paging,
    single_param,
)
from ..schemas import Message, MessageBundle, MessageInput, MessagePage, body_payload
from ..time import now_utc

router = APIRouter()


@router.get("/api/v1/messages", response_model=MessagePage, response_model_exclude_none=True)
def list_messages(request: Request, _caller: OptionalCaller) -> Response:
    page, size = require_valid_paging(request)
    key = single_param(request, "key")
    language = single_param(request, "language")
    q = single_param(request, "q")
    require_max_length(q, 200, "q")
    conditions = [sql.SQL("true")]
    params: tuple[Any, ...] = ()
    if key is not None:
        conditions.append(sql.SQL("key = %s"))
        params = (*params, key)
    if language is not None:
        conditions.append(sql.SQL("language = %s"))
        params = (*params, language)
    if q is not None and q.strip():
        conditions.append(sql.SQL("(key ilike %s escape E'\\\\' or text ilike %s escape E'\\\\')"))
        pattern = f"%{escape_like(q)}%"
        params = (*params, pattern, pattern)
    where = where_clause(conditions)
    rows = query(
        sql.SQL("select * from messages where {where} order by key, language limit %s offset %s").format(where=where),
        (*params, size, page * size),
    )
    total = int(
        one(sql.SQL("select count(*)::int as count from messages where {where}").format(where=where), params)["count"]
    )
    return response_with_etag(request, page_of(rows, page, size, total, to_message_response))


@router.get("/api/v1/messages/bundle", response_model=MessageBundle)
def get_bundle(request: Request, _caller: OptionalCaller) -> Response:
    language = resolve_language(first_param(request, "lang"), request.headers.get("accept-language"))
    return response_with_etag(
        request,
        {"language": language, "messages": message_bundle(language)},
        {"Content-Language": language},
    )


@router.get("/api/v1/messages/{message_id}", response_model=Message, response_model_exclude_none=True)
def get_message(request: Request, message_id: str, _caller: OptionalCaller) -> Response:
    message = one("select * from messages where id = %s", (parse_uuid(message_id),))
    if message is None:
        raise NotFoundProblem()
    return response_with_etag(request, to_message_response(message))


@router.post("/api/v1/messages", status_code=201, response_model=Message, response_model_exclude_none=True)
def create_message(response: Response, caller: AdminCaller, body: MessageInput | None) -> dict[str, Any]:
    input_data = validate_message_input(body_payload(body))
    message_id = str(uuid4())
    now = now_utc()
    with transaction() as conn:
        duplicate = conn.execute(
            "select 1 from messages where key = %s and language = %s",
            (input_data["key"], input_data["language"]),
        ).fetchone()
        if duplicate:
            raise duplicate_message_conflict(input_data)
        try:
            row = conn.execute(
                """
                insert into messages (id, key, language, text, description, created_at, updated_at)
                values (%s, %s, %s, %s, %s, %s, %s)
                returning *
                """,
                (
                    message_id,
                    input_data["key"],
                    input_data["language"],
                    input_data["text"],
                    input_data["description"],
                    now,
                    now,
                ),
            ).fetchone()
        except UniqueViolation as exc:
            raise duplicate_message_conflict(input_data) from exc
        record_audit(conn, caller.username, "message.created", "message", str(row["id"]), message_snapshot(row))
    log_message_event("message_created", "Message created", caller.username, row)
    response.headers["Location"] = f"/api/v1/messages/{message_id}"
    return to_message_response(row)


@router.put("/api/v1/messages/{message_id}", response_model=Message, response_model_exclude_none=True)
def update_message(message_id: str, caller: AdminCaller, body: MessageInput | None) -> dict[str, Any]:
    parsed_message_id = parse_uuid(message_id)
    input_data = validate_message_input(body_payload(body))
    with transaction() as conn:
        existing = conn.execute("select 1 from messages where id = %s", (parsed_message_id,)).fetchone()
        if existing is None:
            raise NotFoundProblem()
        duplicate = conn.execute(
            "select 1 from messages where key = %s and language = %s and id <> %s",
            (input_data["key"], input_data["language"], parsed_message_id),
        ).fetchone()
        if duplicate:
            raise duplicate_message_conflict(input_data)
        try:
            row = conn.execute(
                """
                update messages
                set key = %s, language = %s, text = %s, description = %s, updated_at = %s
                where id = %s
                returning *
                """,
                (
                    input_data["key"],
                    input_data["language"],
                    input_data["text"],
                    input_data["description"],
                    now_utc(),
                    parsed_message_id,
                ),
            ).fetchone()
        except UniqueViolation as exc:
            raise duplicate_message_conflict(input_data) from exc
        record_audit(conn, caller.username, "message.updated", "message", str(row["id"]), message_snapshot(row))
    log_message_event("message_updated", "Message updated", caller.username, row)
    return to_message_response(row)


@router.delete("/api/v1/messages/{message_id}", status_code=204)
def delete_message(message_id: str, caller: AdminCaller) -> Response:
    parsed_message_id = parse_uuid(message_id)
    with transaction() as conn:
        row = conn.execute("delete from messages where id = %s returning *", (parsed_message_id,)).fetchone()
        if row is None:
            raise NotFoundProblem()
        record_audit(conn, caller.username, "message.deleted", "message", str(row["id"]), message_snapshot(row))
    log_message_event("message_deleted", "Message deleted", caller.username, row)
    return Response(status_code=204)

File: backends/node-ts/src/routes/messages.ts
import { randomUUID } from "node:crypto";
import type { FastifyInstance, FastifyRequest } from "fastify";
import { pool, withTransaction } from "../db.js";
import { requireRole } from "../auth.js";
import { recordAudit } from "../audit.js";
import { logEvent } from "../logging.js";
import { sendWithEtag } from "../etag.js";
import { DEFAULT_LANGUAGE, messageBundle, resolveLanguage } from "../i18n.js";
import {
  ConflictProblem,
  NotFoundProblem,
  Validator,
  escapeLike,
  firstParam,
  omitNulls,
  parseUuid,
  requireMaxLength,
  requireValidPaging,
  singleParam,
} from "../problems.js";

const KEY_PATTERN = /^[a-z0-9-]+(\.[a-z0-9-]+)*$/;
const LANGUAGE_PATTERN = /^[a-z]{2}$/;

interface MessageRow {
  id: string;
  key: string;
  language: string;
  text: string;
  description: string | null;
  created_at: Date;
  updated_at: Date;
}

const toMessageResponse = (row: MessageRow) =>
  omitNulls({
    id: row.id,
    key: row.key,
    language: row.language,
    text: row.text,
    description: row.description,
    createdAt: row.created_at.toISOString(),
    updatedAt: row.updated_at.toISOString(),
  });

interface MessageInput {
  key: string;
  language: string;
  text: string;
  description: string | null;
}

function validateMessageInput(body: unknown): MessageInput {
  const input = (typeof body === "object" && body !== null ? body : {}) as Record<string, unknown>;
  const validator = new Validator();

  const key = typeof input["key"] === "string" ? input["key"].trim() : "";
  validator.check(KEY_PATTERN.test(key) && key.length <= 150, "key", "validation.message.key.invalid");

  const language = typeof input["language"] === "string" ? input["language"].trim() : "";
  validator.check(LANGUAGE_PATTERN.test(language), "language", "validation.message.language.invalid");

  const text = typeof input["text"] === "string" ? input["text"] : "";
  validator.check(text !== "", "text", "validation.message.text.required");
  validator.check(text.length <= 2000, "text", "validation.message.text.too-long");

  const description = typeof input["description"] === "string" ? input["description"] : null;
  validator.check((description?.length ?? 0) <= 1000, "description", "validation.message.description.too-long");

  validator.throwIfInvalid();
  return { key, language, text, description };
}

/** The message key is safe to log: validated against `KEY_PATTERN`, so no free-form client text. */
function logMessageEvent(event: string, description: string, actor: string, message: MessageRow): void {
  logEvent("info", event, "success", description, {
    actor,
    resource_type: "message",
    resource_id: message.id,
    message_key: message.key,
    language: message.language,
  });
}

async function requestLanguage(request: FastifyRequest): Promise<string> {
  return resolveLanguage(
    firstParam((request.query as Record<string, unknown>)["lang"]),
    request.headers["accept-language"],
  );
}

export function registerMessageRoutes(app: FastifyInstance): void {
  /** Reads are public with ETag revalidation (SPEC rules 7 + 10). */
  app.get("/api/v1/messages", async (request, reply) => {
    const query = request.query as Record<string, unknown>;
    const { page, size } = requireValidPaging(query);
    const key = singleParam(query["key"], "key");
    const language = singleParam(query["language"], "language");
    const q = singleParam(query["q"], "q");
    requireMaxLength(q, 200, "q");

    const conditions: string[] = ["true"];
    const params: unknown[] = [];
    const bind = (value: unknown): string => {
      params.push(value);
      return `$${params.length}`;
    };
    if (key !== undefined) conditions.push(`key = ${bind(key)}`);
    if (language !== undefined) conditions.push(`language = ${bind(language)}`);
    if (q !== undefined && q.trim() !== "") {
      const pattern = bind(`%${escapeLike(q)}%`);
      conditions.push(`(key ilike ${pattern} escape '\\' or text ilike ${pattern} escape '\\')`);
    }
    const where = conditions.join(" and ");

    // ordered so identical reads produce identical bodies — the ETag depends on it
    const items = await pool.query(
      `select * from messages where ${where} order by key, language limit ${size} offset ${page * size}`,
      params,
    );
    const total = await pool.query(`select count(*)::int as count from messages where ${where}`, params);
    const totalItems = (total.rows[0] as { count: number }).count;
    return sendWithEtag(request, reply, {
      items: (items.rows as MessageRow[]).map(toMessageResponse),
      page,
      size,
      totalItems,
      totalPages: Math.ceil(totalItems / size),
    });
  });

  /** The flat key → text bundle frontends load on startup (SPEC rule 9). */
  app.get("/api/v1/messages/bundle", async (request, reply) => {
    const language = await requestLanguage(request);
    reply.header("content-language", language);
    return sendWithEtag(request, reply, { language, messages: await messageBundle(language) });
  });

  app.get("/api/v1/messages/:id", async (request, reply) => {
    const id = parseUuid((request.params as { id: string }).id);
    const result = await pool.query("select * from messages where id = $1", [id]);
    const message = result.rows[0] as MessageRow | undefined;
    if (!message) throw new NotFoundProblem();
    return sendWithEtag(request, reply, toMessageResponse(message));
  });

  app.post("/api/v1/messages", async (request, reply) => {
    const caller = requireRole(request, "admin");
    const input = validateMessageInput(request.body);
    // mutation + audit commit together (SPEC rule 18: every backoffice mutation is audited)
    const message = await withTransaction(async (client) => {
      const duplicate = await client.query("select 1 from messages where key = $1 and language = $2", [
        input.key,
        input.language,
      ]);
      if (duplicate.rowCount) throw duplicateConflict(input);
      let row: MessageRow;
      try {
        const result = await client.query(
          `insert into messages (id, key, language, text, description, created_at, updated_at)
           values ($1, $2, $3, $4, $5, $6, $6) returning *`,
          [randomUUID(), input.key, input.language, input.text, input.description, new Date()],
        );
        row = result.rows[0] as MessageRow;
      } catch (error) {
        // lost the (key, language) race against a concurrent create — same 409 as the pre-check
        if ((error as { code?: string }).code === "23505") throw duplicateConflict(input);
        throw error;
      }
      await recordAudit(client, caller.username, "message.created", "message", row.id, snapshot(row));
      return row;
    });
    logMessageEvent("message_created", "Message created", caller.username, message);
    return reply.code(201).header("location", `/api/v1/messages/${message.id}`).send(toMessageResponse(message));
  });

  app.put("/api/v1/messages/:id", async (request) => {
    const caller = requireRole(request, "admin");
    const id = parseUuid((request.params as { id: string }).id);
    const input = validateMessageInput(request.body);
    const message = await withTransaction(async (client) => {
      const existing = await client.query("select 1 from messages where id = $1", [id]);
      if (!existing.rowCount) throw new NotFoundProblem();
      const duplicate = await client.query("select 1 from messages where key = $1 and language = $2 and id <> $3", [
        input.key,
        input.language,
        id,
      ]);
      if (duplicate.rowCount) throw duplicateConflict(input);
      let row: MessageRow;
      try {
        const result = await client.query(
          `update messages set key = $2, language = $3, text = $4, description = $5, updated_at = $6
           where id = $1 returning *`,
          [id, input.key, input.language, input.text, input.description, new Date()],
        );
        row = result.rows[0] as MessageRow;
      } catch (error) {
        if ((error as { code?: string }).code === "23505") throw duplicateConflict(input);
        throw error;
      }
      await recordAudit(client, caller.username, "message.updated", "message", row.id, snapshot(row));
      return row;
    });
    logMessageEvent("message_updated", "Message updated", caller.username, message);
    return toMessageResponse(message);
  });

  app.delete("/api/v1/messages/:id", async (request, reply) => {
    const caller = requireRole(request, "admin");
    const id = parseUuid((request.params as { id: string }).id);
    const message = await withTransaction(async (client) => {
      const result = await client.query("delete from messages where id = $1 returning *", [id]);
      const row = result.rows[0] as MessageRow | undefined;
      if (!row) throw new NotFoundProblem();
      await recordAudit(client, caller.username, "message.deleted", "message", row.id, snapshot(row));
      return row;
    });
    logMessageEvent("message_deleted", "Message deleted", caller.username, message);
    return reply.code(204).send();
  });
}

const duplicateConflict = (input: { key: string; language: string }): ConflictProblem =>
  new ConflictProblem(`A message with key '${input.key}' and language '${input.language}' already exists.`);

const snapshot = (message: MessageRow): Record<string, unknown> => ({
  key: message.key,
  language: message.language,
  text: message.text,
  description: message.description,
});

export { DEFAULT_LANGUAGE };
